fairscale/nn/misc/checkpoint_activations.py

```diff
@@ -41,8 +40,23 @@ def checkpoint_wrapper(module: nn.Module, offload_to_cpu: bool = False) -> nn.Mo
         (nn.Module):
             wrapped module
     """
-    module.forward = functools.partial(_checkpointed_forward, module.forward, offload_to_cpu)  # type: ignore
-    return module
+    # Do not use functools.partial like:
+    #
+    #     module.forward = functools.partial(_checkpointed_forward, module.forward, offload_to_cpu)
+    #
+    # It causes the backward to hold-on to tensor memory even when model is
+    # freed.
+
+    # Use a wrapper to wrap the original module.
+    class CheckpointWrapper(nn.Module):
```

Comment:
One downside of this is that it doesn't preserve `state_dict`, so models that are wrapped with the checkpoint wrapper can no longer be loaded without the checkpoint (and vice-versa).

A better approach might be what we do in FlattenParamsWrapper: https://github.com/facebookresearch/fairscale/blob/master/fairscale/nn/misc/flatten_params_wrapper.py#L126-L136
And corresponding test: https://github.com/facebookresearch/fairscale/blob/master/tests/nn/misc/test_flatten_params_wrapper.py#L122-L129